### DISCUSSION TITLE
Accessing GtkWidget class methods

### DISCUSSION THREAD

**Discussion by danirod** [gtk-rs/gtk4-rs#1790] [Q&A]
Hello there. I am looking for suggestions in regards to using [the class methods for the GtkWidget class](https://docs.gtk.org/gtk4/class.Widget.html#class-methods). In gtk-rs, these methods are part of the [WidgetClassExt](https://docs.rs/gtk4/latest/gtk4/subclass/widget/trait.WidgetClassExt.html), but I can't get my head around how to get access to those methods. In my case, the goal is to have access to the `add_shortcut` method ([docs.rs](https://docs.rs/gtk4/latest/gtk4/subclass/widget/trait.WidgetClassExt.html#method.add_shortcut), [GTK](https://docs.gtk.org/gtk4/class_method.Widget.add_shortcut.html)) so that I can install shortcuts into every instance of a specific widget.

I am writing a GTK 4 application and while GNU/Linux is the preferred platform, I am doing my best to make sure that the application looks and works right in Windows and macOS too. In GTK 4, the `<Primary>` keyboard modifier is bound to the Control key even on macOS, where most users would expect the `<Meta>` modifier. This means that, by default, the shortcuts for built-in actions like copy and paste from a text field are `Ctrl-C` and `Ctrl-V` rather than `Cmd-C` and `Cmd-V`, unless the application changes them.

[In this specific example](https://gaphor.org/2022/12/10/gtk4-macos-keybindings/#built-in-widgets), Gaphor is using the following Python code to change the shortcuts of the selection.select-all action part of the Gtk.TextView widget class, so that every instance of Gtk.TextView uses the new shortcut (the code is actually more complex becuase they are using a function to encapsulate building the shortcut):

```python
Gtk.TextView.add_shortcut(
    new_shortcut_with_args(
        "<Meta>a", "select-all",
        GLib.Variant.new_boolean(True)
    )
)
```

One of the benefits of the GObject introspection is that it is very easy to translate code from between languages. However, even if I suspect it cannot be that difficult, I can't call the `add_shortcut` function because I don't know how to get a code path to that method.

If this was a custom widget, I could just use the `class_init` function when implementing `ObjectSubclass` for my widget, because I have a Class parameter, but when I try to apply it to the GtkText and GtkTextView widgets like the code I show above, I am not having any luck.

For C and Python, the function calls are more obvious (`gtk_widget_class_add_shortcut` and `Gtk.TextView.add_shortcut`), but in Rust it is being a challenge. I don't know if I am supposed to just use `gtk::TextView::add_shortcut()` and make sure that I am using the WidgetClassExt trait in my file (on a first try, doesn't seem to work), or if I have to use a function to get the class first. I've played with the following, but it doesn't compile.

```rust
let trigger = gtk::ShortcutTrigger::parse_string("<Meta>c").unwrap();
let shortcut = gtk::Shortcut::new(Some(trigger), Some(gtk::NamedAction::new("clipboard.copy")));

let text_type = gtk::TextView::static_type();
let text_class = glib::object::Class::from_type(text_type).unwrap();
text_class.add_shortcut(&trigger);
```

I've checked the source code for other programs written in gtk-rs to see if I can learn something from them, but it seems that none of them has had the need to use this function so far.

Any advice? Thanks in advance.

**Comment by bilelmoussaoui** [comment]
The problem comes from the trait constraints imposed at 
```rust
unsafe impl<T: ClassStruct> WidgetClassExt for T where T::Type: WidgetImpl {}
```
it should be something like 
```rust
unsafe impl<T: IsA<Widget> + glib::object::IsClass> WidgetClassExt for glib::object::Class<T> {}
```
instead

**Reply by bilelmoussaoui** [reply]
Could you convert the discussion to an issue please?

**Reply by danirod** [reply]
Sure!

Created issue as https://github.com/gtk-rs/gtk4-rs/issues/1793

**Comment by grewalrehmat** [comment]
I spent some time digging into #2345 because at first I thought the issue was that the `WidgetClass` methods themselves were missing or broken in gtk-rs. After tracing the implementation and git history, it looks like the main confusion is around how the class methods are exposed and what the two different APIs actually expect.

Originally, these methods were exposed through `WidgetClassSubclassExt`. That API is built around `ClassStruct`, so it makes sense when you're inside the `class_init` of a custom widget where you already have the class parameter. It doesn't really help when you're dealing with an existing GTK type like `TextView` and want to modify its class.

The newer `WidgetClassManualExt` implementation from `d0f606a09d` / `e38a33655d` was added to solve that general problem. It exposes the class methods through `glib::Class<T>` for widget types, so the intended path is to get a `Class<TextView>` and call `add_shortcut()` on it.

The more interesting part for #2345 is the type mismatch. `WidgetClassManualExt` is implemented for `glib::Class<T>` only when `T` satisfies the required `IsA<Widget> + IsClass` bounds. So the concrete `T` matters, and with the way `Class::from_type()` is used, the compiler can end up not having enough information to infer the `T` needed for the trait implementation. The underlying GTK class is valid, but Rust basically goes "yeah, but which `T` tho" and refuses to resolve the method.

But, as suggested by @sdroege, I don't think simply reverting those commits would be a good fix. They were fixing an actual limitation: before #1793, these class methods were pretty much locked behind the subclassing API, so working with an existing GTK widget like `TextView` wasn't really possible through the same path. Reverting #1793 would just bring that limitation back instead of actually fixing #2345.
